### PR TITLE
[V1][Spec Decode] Handle draft tokens beyond max_model_len

### PR DESCRIPTION
**Pull request by WoosukKwon**
Implements 4. Handle the edge cases like when the draft model generates beyond max_pos_embeddings in #15901

### CODE REVIEW

**Review by comaniac — APPROVED** [review]
Overall LGTM. Approve first to unblock the PR. Meanwhile it would be good to have a unit test for it. Also do we know the overhead of introduced ops (e.g., torch.where)?

File: vllm/v1/spec_decode/eagle.py [hunk 1]
```diff
@@ -20,7 +22,10 @@ def __init__(
         self.vllm_config = vllm_config
         self.num_speculative_tokens = (
             vllm_config.speculative_config.num_speculative_tokens)
+        self.max_model_len = vllm_config.model_config.max_model_len
         self.block_size = vllm_config.cache_config.block_size
+
+        # Optimization: pre-compute and cache the arange tensor.
```
Comment (ekagra-ranjan) [inline vllm/v1/spec_decode/eagle.py:28]: maybe we could also pass in one used here: https://github.com/vllm-project/vllm/blob/027b204ff15e803190775e04d39973606a3a7021/vllm/v1/worker/gpu_model_runner.py#L240

File: vllm/v1/spec_decode/eagle.py [hunk 2]
```diff
@@ -103,22 +108,52 @@ def propose(
...
-                                          positions % self.block_size)
+                                          clamped_positions % self.block_size)
+            # Mask out the slot mappings that exceed the max model length.
+            # Otherwise, the KV cache will be inadvertently updated with the
+            # padding tokens.
+            attn_metadata.slot_mapping = torch.where(
+                exceeds_max_model_len,
+                PADDING_SLOT_ID,
+                attn_metadata.slot_mapping,
+            )
```
Comment (ekagra-ranjan) [inline vllm/v1/spec_decode/eagle.py:149]: My understanding is that torch.where will allocate an intermedia tensor and then assign it. Is it possible to use `attn_metadata.slot_mapping[exceeds_max_model_len] = PADDING_SLOT_ID` to that it's an in-place operation?

File: vllm/v1/spec_decode/eagle.py [hunk 3]
```diff
@@ -103,22 +108,52 @@ def propose(
             # Update the inputs.
             input_ids = draft_token_ids_list[-1]
             positions += 1
+
+            # NOTE(woosuk): We should handle the case where the draft model
+            # generates tokens beyond the max model length. Since it is complex
+            # to remove such requests from the batch, we keep them in the batch
+            # but adjust the position ids and slot mappings to avoid the
+            # out-of-range access during the model execution. The draft tokens
+            # generated with this adjustment should be ignored.
```
Comment (ekagra-ranjan) [inline vllm/v1/spec_decode/eagle.py:117]: Do we need to add the logic to ignoring such draft tokens [here](https://github.com/vllm-project/vllm/pull/16087/files#diff-80ee7e2a62f9dcfbb8a312dc4e3948557e97ef187290daebbcae1e28596bda29R1200)?

**Comment by WoosukKwon** [comment]
@comaniac Good point. Added a test. Also, I replaced `torch.where` with `masked_fill_` wherever possible, for better performance. Overall, I think the overhead will be very small because the masked tensors are all small (i.e., shape of `[batch_size]`).

File: vllm/v1/spec_decode/eagle.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py]: @ekagra-ranjan Thanks for the suggestion. I changed it to `masked_fill_`, which is also an in-place operation.
Overall, I think the performance impact will be small since the tensors here are small (shape of `[batch_size]`).

File: vllm/v1/spec_decode/eagle.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:124]: Good question. The scheduler handles it https://github.com/vllm-project/vllm/blob/af7462b6d1ebc10ca83da57f2804c9e49a40e83c/vllm/v1/core/sched/scheduler.py#L188-L193
